### PR TITLE
Fix #8423: Remove undefined while getting the type of the first argument of then signature

### LINKED ISSUE #8423: Regression: TypeScript no longer awaits custom Promises
**TypeScript Version:** 

1.9.0-dev.20160502

**Code**

```ts
declare namespace Windows.Foundation {
    interface IPromise<TResult> {
        then<U>(success?: (value: TResult) => IPromise<U>, error?: (error: any) => IPromise<U>, progress?: (progress: any) => void): IPromise<U>;
        then<U>(success?: (value: TResult) => IPromise<U>, error?: (error: any) => U, progress?: (progress: any) => void): IPromise<U>;
        then<U>(success?: (value: TResult) => U, error?: (error: any) => IPromise<U>, progress?: (progress: any) => void): IPromise<U>;
        then<U>(success?: (value: TResult) => U, error?: (error: any) => U, progress?: (progress: any) => void): IPromise<U>;
        done<U>(success?: (value: TResult) => any, error?: (error: any) => any, progress?: (progress: any) => void): void;

        cancel(): void;
    }
}

async function sample(promise: Windows.Foundation.IPromise<number>) {
    var number = await promise;
}
```

**Expected behavior:**
No error, as this is an await-compatible Promise definition; more concretely, those are definitions for UWP APIs and they worked well with a stable TypeScript.

**Actual behavior:** 
Compiler emits an error `Operand for 'await' does not have a valid callable 'then' member.`.

### PR DESCRIPTION
**Pull request by mhegazy**
Fixes #8423

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -13443,7 +13443,7 @@ namespace ts {
         }
 
         function getTypeOfFirstParameterOfSignature(signature: Signature) {
-            return getTypeAtPosition(signature, 0);
+            return getTypeWithFacts(getTypeAtPosition(signature, 0), TypeFacts.NEUndefined);
```
Comment (vladima) [inline src/compiler/checker.ts]: this function is used in two places 
- when we get type of first parameter in the call signature of `then`
- when we get type of first parameter in the `onFulfilled` callback. 

Do we want to strip `undefined` in both places or just in the first one?

Comment (mhegazy) [inline src/compiler/checker.ts]: well. i went back and forth on this, and then did not see a reason why not... 
so we want to get `T`

`then<U>(success?: (value: T) => U,...): IPromise<U>`

the first takes care of the optionality of `succes?`, the second would remove the `undefined` if `value` was optional. the question is what is the promised type for this:

`then<U>(success?: (value?: T) => U,...): IPromise<U>`

is it `T` or `T|undefined` ?

Comment (vladima) [inline src/compiler/checker.ts]: what about cases like this:
```ts
declare function resolve1<T>(value: T): Promise<T>;
declare function resolve2<T>(value: T): Windows.Foundation.IPromise<T>;

async function f(x?: number) {
    let x1 = await resolve1(x);
    let x2 = await resolve2(x);
}
```
Currently type of `x1` is `number | undefined` and `x2` fails with error addressed by this PR.
With the fix type of `x2` will be just `number` which will be inconsistent with type for `x1`

Comment (mhegazy) [inline src/compiler/checker.ts]: good point. fixed. thanks

**Comment by vladima** [comment]
:+1:

**Pull request merged**